### PR TITLE
Download rules and unarchive them

### PR DESCRIPTION
**Pull request by pbalogh-sa**

### CODE REVIEW

File: pkg/rules/rules.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+		if !prevStepFailed {
+			logger.Infof("Unarchive rules file=%s into %s", fileName, tmpSourceDir)
+			if err := unzip(fileName, tmpSourceDir); err != nil {
+				logger.Errorf("Failed to unarchive source %s. Using last successful download if available. File=%s: %v", source.Name, fileName, err)
+				prevStepFailed = true
+			}
+		}
+
+		// Replace contents of source dir with the downloaded and unarchived data
+		if !prevStepFailed {
```
Comment (Tehsmash) [inline pkg/rules/rules.go:78]: Not sure this is a sensible code structure, I suggest that we move this out into a separate function like:
```
// atomicDownloadAndReplace will download and unarchive the URL,
// only if those are successful will the data replace sourceDir
func atomicDownloadAndReplace(url, sourceDir) error {
   tmpdir, err := os.MkdirTemp(...)
   if err != nil {
       return fmt.Errorf(...., err)
   }

   err := downloadFile(...)
   if err != nil {
       return ...
   }
   
   err := unzip(...)
   if err != nil {
      return ...
   }
   
   err := os.RemoveAll(...)
   if err != nil {
      return ...
   }
   
   err := os.Rename(...)
   if err != nil {
      return ...
   }
}   
```

In this function it simple becomes:

```
		sourceDir := path.Join(config.CacheDir, "sources", source.Name)
		if err := os.MkdirAll(sourceDir, 0755); err != nil {
			logger.Errorf("failed to create directory for source %s: %v", sourceDir, err)
			continue
		}
		
		err := atomicDownloadAndReplace(url, sourceDir)
		if err != nil {
           logger.Errorf("unable to update source %s; using any existing data: %v", err)
		}
		
		yarFiles, err := createYarFileListToIndex(sourceDir, reg)
		if err != nil {
		   logger.Errorf(...)
		   continue
		}
```

**Review by Tehsmash — CHANGES_REQUESTED** [review]

File: pkg/rules/rules.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+	"github.com/openclarity/yara-rule-server/pkg/config"
+)
+
+func DownloadAndCompile(cfg *config.Config, logger *logrus.Entry) error {
+	// First try to download new copies of all the sources
+	yarFilesToIndex := make([]string, 0)
+	tempDir := path.Join(config.CacheDir, "tmp")
+	if err := os.MkdirAll(tempDir, 0755); err != nil {
+		return fmt.Errorf("failed to create yara rule server temp directory. tempDir=%s error=%v", tempDir, err)
```
Comment (FrimIdan) [inline pkg/rules/rules.go:37]: no need the `error=%v` we usually use only `:%w`, right?

File: pkg/rules/rules.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+	}
+	for _, source := range cfg.RuleSources {
+		prevStepFailed := false
+		// Create directory for this source if it doesn't exist
+		sourceDir := path.Join(config.CacheDir, "sources", source.Name)
+		if err := os.MkdirAll(sourceDir, 0755); err != nil {
+			logger.Errorf("failed to create directory %s. Using the last successful download if available: error=%v", sourceDir, err)
```
Comment (FrimIdan) [inline pkg/rules/rules.go:44]: no need the `error=%v` we usually use only `:%w`, right?

File: pkg/rules/rules.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+			prevStepFailed = true
+		}
+
+		// Download the source URL.
+		//tmpSourceDir, err := os.MkdirTemp(tempDir, source.Name+"-yara-rule")
+		var tmpSourceDir string
+		var err error
+		if !prevStepFailed {
+			if tmpSourceDir, err = os.MkdirTemp(tempDir, source.Name+"-yara-rule"); err != nil {
+				logger.Errorf("failed to create temp directory for %s Using the last successful download if available: error=%v", source.Name, err)
```
Comment (FrimIdan) [inline pkg/rules/rules.go:54]: please remove all the `error` from the error prints

File: pkg/rules/rules.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+	yarFilesToIndex := make([]string, 0)
+	tempDir := path.Join(config.CacheDir, "tmp")
+	if err := os.MkdirAll(tempDir, 0755); err != nil {
+		return fmt.Errorf("failed to create yara rule server temp directory. tempDir=%s: %w", tempDir, err)
+	}
+	for _, source := range cfg.RuleSources {
+		// Create directory for this source if it doesn't exist
+		sourceDir := path.Join(config.CacheDir, "sources", source.Name)
+		if err := os.MkdirAll(sourceDir, 0755); err != nil {
+			logger.Errorf("failed to create directory %s. Using the last successful download if available: %w", sourceDir, err)
```
Comment (Tehsmash) [inline pkg/rules/rules.go:43]: As the linter said, %w is only for fmt.Errorf. When logging errors we should use:
```
logger.WithError(err).Errorf("words words words")
```

**Review by Tehsmash — CHANGES_REQUESTED** [review]

File: pkg/rules/rules.go [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+func atomicDownloadAndReplace(source config.RuleSource, sourceDir, tempDir string, logger *logrus.Entry) error {
+	tmpSourceDir, err := os.MkdirTemp(tempDir, source.Name+"-yara-rule")
+	if err != nil {
+		return fmt.Errorf("failed to create temp directory for %s Using the last successful download if available: %w", source.Name, err)
+	}
+
+	fileName := filepath.Join(tmpSourceDir, source.Name+".zip")
+	logger.Infof("Downloading %s into %s", source.URL, fileName)
+	if err := downloadFile(fileName, source.URL); err != nil {
+		return fmt.Errorf("failed to download rule source %s. Using the last successful download if available. URL=%s: %v", source.Name, source.URL, err)
```
Comment (Tehsmash) [inline pkg/rules/rules.go:102]: should be %w for wrapping the error

File: pkg/rules/rules.go [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+	}
+
+	logger.Infof("Unarchive rules file=%s into %s", fileName, tmpSourceDir)
+	if err := unzip(fileName, tmpSourceDir); err != nil {
+		return fmt.Errorf("failed to unarchive source %s. Using last successful download if available. File=%s: %v", source.Name, fileName, err)
```
Comment (Tehsmash) [inline pkg/rules/rules.go:107]: ditto

File: pkg/rules/rules.go [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+	}
+
+	// Replace contents of source dir with the downloaded and unarchived data
+	if err := os.RemoveAll(sourceDir); err != nil {
+		return fmt.Errorf("failed to remove previous sources: %v", err)
```
Comment (Tehsmash) [inline pkg/rules/rules.go:112]: ditto

File: pkg/rules/rules.go [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+	}
+	if err := os.Rename(tmpSourceDir, sourceDir); err != nil {
+		return fmt.Errorf("failed to move downloaded source: %v", err)
```
Comment (Tehsmash) [inline pkg/rules/rules.go:115]: ditto

**Review by Tehsmash — CHANGES_REQUESTED** [review]

File: yara-rule-server-example [hunk 10]
```diff
@@ -0,0 +1,9 @@
+enable_json_log: true
+rule_update_schedule: "0 0 * * *"
+rule_sources:
+  - name: "base"
+    url: "https://github.com/Yara-Rules/rules/archive/refs/heads/master.zip"
+    exclude_regex: ".*index.*.yar|.*/utils/.*|.*/deprecated/.*|.*index_.*|.*MALW_AZORULT.yar"
```
[the remark this answers is not in the document]
Comment (pbalogh-sa) [reply yara-rule-server-example]: I've opened an issue: https://github.com/openclarity/yara-rule-server/issues/6

**Review by Tehsmash — APPROVED** [review]
LGTM, there are a few more places which need the error wrapping fixed.

We can copy this https://github.com/openclarity/vmclarity/pull/668 and enable the linter to check this and fix it in a follow up.

File: pkg/rules/compile.go [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+	"github.com/sirupsen/logrus"
+)
+
+func generateIndex(indexGenPATH string, logger *logrus.Entry) error {
+	indexGen := exec.Command(indexGenPATH)
+
+	resultJsonB, err := indexGen.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("failed to run index_gen.sh: %v, %v", err, string(resultJsonB))
```
[the remark this answers is not in the document]
Comment (pbalogh-sa) [reply pkg/rules/compile.go]: index_gen.sh is not used anymore

File: pkg/config/config.go [hunk 12]
```diff
@@ -23,8 +23,15 @@ import (
...
 
 type Config struct {
-	LogLevel      string `mapstructure:"LOG_LEVEL"`
-	EnableJSONLog bool   `mapstructure:"ENABLE_JSON_LOG"`
+	LogLevel                  string   `mapstructure:"log_level"`
+	EnableJSONLog             bool     `mapstructure:"enable_json_log"`
+	RulePath                  string   `mapstructure:"rule_path"`
+	RuleURLs                  []string `mapstructure:"rule_urls"`
+	YaracPath                 string   `mapstructure:"yarac_path"`
+	IndexGenPath              string   `mapstructure:"index_gen_path"`
```
[the remark this answers is not in the document]
Comment (pbalogh-sa) [reply pkg/config/config.go]: index_gen.sh is not used anymore

**Review by akpsgit — APPROVED** [review]

**Pull request merged**